### PR TITLE
Refactor Dockerfiles with best practices

### PR DESCRIPTION
**Pull request by PeterDaveHello**
Implement some best practices like: prevent leaving and remove unnecessary cache, also remove some unnecessary commands.

### CODE REVIEW

**Comment by thc202** [comment]
What are the size and time differences?

**Comment by PeterDaveHello** [comment]
Won't be too much actually, so I didn't precisely calculate the result, but as you can see, some of them are alreay done in the Dockerfiles, just not all of it done in the proper way, so a quick fix should be helpful.

**Comment by thc202** [comment]
The commits could be squashed, changes are not that complex or disjoint that need to be in their own commits.

IMO the changelog should be updated, even if it's just to mention the Docker files were tidied up.

**Comment by PeterDaveHello** [comment]
@thc202 just want to share the idea is trying to focus every commit to a single and simple topic, which will make sure the commit is atomic, should make sure the change are easier to be understood and traced in the future, especially when something went wrong, but the same type of changes cross different Dockerfiles, will just be in the same commit. Still got your point and understand your idea though!

BTW, should I record the change in a certain file? Not sure if I missed it in the docs...

**Comment by kingthorin** [comment]
Thanks for clarifying your thoughts on the commits. I don't disagree, however, the changes are all related and the individual commit history isn't really relevant to the greater project in the end. So squashing it is best.

Docker changes are tracked in: https://github.com/zaproxy/zaproxy/blob/main/docker/CHANGELOG.md

**Comment by thc202** [comment]
This needs to be rebased, one of the files was removed (not in use).

File: docker/Dockerfile-bare [hunk 1]
```diff
@@ -36,7 +36,6 @@ COPY --link --chown=1000:1000 policies /home/zap/.ZAP/policies/
 RUN echo "zap2docker-bare" > /zap/container
 
 RUN /usr/sbin/adduser -u 1000 -h /home/zap -s /bin/bash -D zap && \
```
Comment (thc202) [inline docker/Dockerfile-bare:38]:
```suggestion
RUN /usr/sbin/adduser -u 1000 -h /home/zap -s /bin/bash -D zap
```

**Comment by thc202** [comment]
Addressed the comments.

**Review by kingthorin — APPROVED** [review]

**Review by psiinon — APPROVED** [review]

**Comment by PeterDaveHello** [comment]
Thanks for help from @thc202, and sorry for the late reply/update 😅

**Comment by thc202** [comment]
Thank you!

**Pull request merged**